### DISCUSSION TITLE
Split Gradle task

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#7674] [Ideas]
Right now we have 3 different tasks on our Gradle Plugin:

- detekt
- detektBaseline
- generateConfig

`generateConfig` is fine. I want to talk about the other two.

Right now `detekt` and `detektBaseline` do nearly the same. The only difference is that one generates a file when there are issues and the other fails. But even doing nearly the same we don't use the "power" of Gradle to cache that part of the tasks that does the same.

This is the main reason to split the `detekt` task in two different tasks.

The first task would generate a file `.detekt` that will contain all the information about the issues found on the code. Then depending in what the user wants we can run the "check" tasks that read that file, generates the reports and fail if there are issues. Or we could run "baseline" that read the `.detekt` file and generates a baseline file based on that.

How can this even help us with the cacheability? My idea is that the "check" task is not cacheable. There are use cases in the reports where you want to show absolute paths. For example, we could have `<a href="file:///path/to/file.kt">` on our html report. The problem of those reports is that they are not "shareable" and for that reason they shouldn't be cacheables. I could even go further and say that this task shouldn't even be "UP-TO-DATE" because it contains console reports like show the number of files/functions on the project that you can only see the first time that you execute the task. But probably this is not a good idea for such an edge case use case.

Loosing cacheability in one task isn't good. I know, but I think that the trade offs are ok. The "check" task will likely be a final task so no one will depend on it. And "check" should be really fast because it just need to read a file and generate other files.

## TL;DR

3 tasks: 
- `detekt` -> generates `.detekt`
- `detektCheck` -> depend on `detekt` and generates reports and fails if there are issues
- `detektBaseline` -> depend on `detekt` and generates the baseline.

`detekt` is fully cacheable because it doesn't need any absolute path. `detektCheck` is not cacheable because some reports want or could want to use absolute paths.

**Comment by cortinico** [comment]
cc @3flex

**Comment by lunaynx** [comment]
Bump, would love to see this (also with `detektMain` <-> `detektBaselineMain`).

**Reply by BraisGabin** [reply]
There is work in progress here. https://github.com/detekt/detekt/pull/8583 is needed to make this happen. And before that one other PR needs to be merged. I want to be clear, #8583 doesn't implement this. But it is a really important milestone to be able to implement this.
